### ISSUE TITLE
Index syntax with keys that are arrays (collections) should prevent assignments

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#18800]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

A follow-up of sorts to https://github.com/PowerShell/PowerShell/issues/18779 (now closed), building on sleuthing by @mikeclayton.

```powershell
$ht = @{ one=1; two=2 }

$arr = 'one', 'two'

'--- Lookup'
$ht[$arr]

'-- Assignment'
$ht[$arr] = 'really?'
$ht
```

### Expected behavior

> `$ht[$arr] = 'really?'`

should cause an error, analogous to how 
* `([pscustomobject] @{ prop=1 }, [pscustomobject] @{ prop=2 }).prop` 
works (get)

but 

* `([pscustomobject] @{ prop=1 }, [pscustomobject] @{ prop=2 }).prop = 42` 
fails (set) (albeit with an unsatisfactory error message) - see https://github.com/PowerShell/PowerShell/issues/5271#issuecomment-340285892

### Actual behavior

```console
--- Lookup
1
2
-- Assignment

Name                           Value
----                           -----
one                            1
two                            2
{one, two}                     really?
```

That is, on _assignment_, the elements of array `$arr` were _not_ considered individually, but the array was used _as a whole_ to create a new entry, which is unexpected.

### Environment data

```powershell
PowerShell Core 7.3.1
```

**Comment by 237dmitry** [comment]
Maybe for some specific tasks, it can be useful?

```powershell
 $ht = @{ 'one' = 1; 'two' = 2 }
 $arr = 'one', 'two'
 
 $ht[$arr] = 3 
 $ht.$arr
```

**Comment by mklement0** [comment]
@237dmitry, the asymmetry between implicit stringification on entry _creation_ and _individual-element lookup_ is just too obscure to be useful in real life; it get's especially obscure if you use `+=`:

```powershell
$ht = @{ 'one' = 1; 'two' = 2 }
$arr = 'one', 'two'
 
# Note the use of +=
$ht[$arr] += 3

$ht.$arr  # !! *array* of values: @(1, 2, 3), resulting from the individual lookups and the assigned value
```

That is, for the purpose of _updating / creating_, the array key was _stringified_. 
But for the purpose of _looking up the entry's existing value_, the array elements were looked up individually.

In fact, with a generic dictionary of `[int]` values, the equivalent of the above code already _does_ break, albeit obscurely:

```
# Note the use of [int] as the value type.
$gd = [System.Collections.Generic.Dictionary[string, int]]::new()
$gd['one'] = 1
$gd['two'] = 2

$arr = 'one', 'two'
 
# Note the use of +=
# !! ERROR:
# !!   Unable to index into an object of type "System.Collections.Generic.Dictionary`2[System.String,System.Int32]"
$gd[$arr] += 3
```

**Comment by 237dmitry** [comment]
> the asymmetry between the implicit string when creating an entry and searching for individual elements is too obscure to be useful in real life.

I don't know how many people have so many thoughts. Maybe someone for something wants to select an array by condition.

```powershell
$ht = @{}
$arr1 = 1, 2
$arr2 = 'three','four'
$arr3 = 5, 'six'   
                   
$ht[$arr1] = $false                  
$ht[$arr2] = $true                   
$ht[$arr3] = $false                  
                                     
$ht.keys.foreach{ if ($ht.$_) { $_ }}
```

**Comment by mklement0** [comment]
Note that array keys don't make much sense by default , since _reference equality_ is tested for on lookup (e.g., the following lookup fails: `$ht = @{ ('one', 'two') = 42 }; $ht.('one', 'two')`; while it _is_ possible to create a dictionary that _meaningfully_ uses array-valued keys, it requires extra work - see [this StackOverflow answer](https://stackoverflow.com/a/72793529/45375) for an example).

With  a generic dictionary and `[string]`-typed keys, as shown in the second snippet in my previous comment, you get _implicit stringification_  on creation / update (e.g. `('one', 'two')` becomes `'one two'`), and element-individual processing on lookup, resulting in the broken behavior shown.

Therefore, I think the best way forward is:

* When using _index_ syntax (`[$arr]`):

  * On _lookup_, _always_ treat arrays / collections as _multiple_ keys to look up _individually_.
  * _Prevent_ this syntax when trying to _assign or update_.

* _Property_-access syntax (`.$arr`) can be used in those rare cases where there are truly array-valued keys.

**Comment by mikeclayton** [comment]
So I don't really have an opinion on what *should* happen, but I know I found the code below to be very "surprising" until I worked it through and realised what it was doing...

```powershell
$sl = [System.Collections.Generic.SortedList[string, array]]::new();

$sl["aaa"] = @(100);
$sl["bbb"] = @(200);

$sl[@("aaa", "bbb")]  = @(300);
$sl[@("aaa", "bbb")] += @(400);

$sl["aaa bbb"]
# 100
# 200
# 400
```

Basically, because of the differences between the way the left-hand side and right-hand side interpreted, this line:

```powershell
$sl[@("aaa", "bbb")] += @(400);
```
is equivalent to 
```powershell
$sl["aaa bbb"] = (@("aaa", "bbb") | % { $sl[$_] }) + @(400);
```

I can see the benefit of the behaviours on the LHS and RHS in isolation, but they combine to do some quite unexpected stuff...

**Comment by mklement0** [comment]
Yes, it's quite confusing, and even once you learn about this counterintuitive asymmetry, it's easy to forget it, and therefore to get tripped up repeatedly over time.

What I've proposed in my previous comment is a simple way to avoid the confusion, without taking away functionality (use property-access syntax or `.Item()` for those rare cases where array keys are meaningfully used).

It is technically a breaking change, but my guess is that it falls into [bucket 3](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area), which would make it acceptable:
* true array keys are presumably rare, because they require non-obvious additional effort to make them work meaningfully
* _incidental_ use of arrays that rely on _implicit stringification_ for use as `[string`] keys (e.g., `@('one', 'two')` -> `'one two'`) is probably also not common (and, arguably, that implicit stringification shouldn't happen to begin with).

The bottom line is:

* While assignments and updates _technically_ accept an array-valued key, the resulting behavior is unexpected with `=`, and downright _broken_ with `+=`.
* That is, no _useful_ functionality is taken away if such assignments and updates are disabled.

**Comment by mklement0** [comment]
In terms of **guidance**, which is worth providing via documentation either way:

* PowerShell's index syntax `[...]` is more than just an alias of the underlying type's parameterized `.Item` property: it supports _arrays_ of indices, which allows retrieving _multiple_ elements from a collection, including from arrays and hashtables (dictionaries), the latter using _key values_ as indices, which are usually strings.

  * Multiple indices are usually expressed as array _literals_, but expressions or variable references that evaluate to / contain arrays work the same; e.g.:

```powershell
@(1..5)[0, -1]  # -> @(1, 5)

$indices = @(0, -1); @(1..5)[$indices]  # ditto
```

* You can use this technique only for _retrieval_ of elements; trying to _assign_ to the retrieved elements usually fails:

```powershell
# -> ERROR: "Array assignment to [0,-1] failed because assignment to slices is not supported."
@(1..5)[0, -1]  = 42 
```

* However, due to an inconsistency, multiple indices (keys) passed to hashtables (dictionaries) _as expressions or via variable values_ are unexpectedly used _as-is, as a whole_ as keys when entries are _created_ or _updated_.

```powershell
# OK - assignment is *prevented*, though the error message is obscure:
#  -> ERROR: 'Unable to index into an object of type "System.Collections.Hashtable".'
 (@{})['one', 'two'] = 3

# !! Unexpectedly, passing the same array as an expression or via a variable *succeeds* and
# !! creates an entry that uses the array as a whole, i.e. @('one', 'two'):
# Note that $ht[$arr] does NOT work for retrieving the entry later, because the normal behavior then applies:
# The array's elements are looked up *individually*.
$arr = @('one', 'two'); ($ht=@{})[$arr] = 3  
```

* In the rare event that you have dictionaries with array-valued (collection-valued) keys or, use _property access syntax_ instead.

```powershell
$arr = @('one', 'two')
$ht = @{}
# Note the use of "." (property access syntax) rather than [...] (index syntax):
# Note that by default only the very same array instance that was used to create the entry can be used
# to retrieve it later.
# To overcome this restriction, define the hashtable as follows, for instance (case-sensitive):
#    $ht = [hashtable]::new([Collections.StructuralComparisons]::StructuralEqualityComparer)
$ht.$arr = 3    # create entry with array key
$ht.$arr        # retrieve entry with array key -> 3
```

* In the similarly rare event that  you want to have array values coerced to a string (e.g., `@('one', 'two')` -> `'one two'`)  with `[string]`-typed keys of generic dictionaries such as `[System.Collections.Generic.Dictionary[string, object]]`,  you can stick with index syntax if you _explicitly_ stringify:

```powershell
# Example with strongly typed [string] keys that array arguments are coerced to:
$arr = @('one', 'two')
$dict = [System.Collections.Generic.Dictionary[string, int]]::new()

# Index syntax: use explicit stringification:
$dict["$arr"] = 3  # create entry with coerced-to-string key 'one two'
$dict["$arr"]      # retrieve entry with coerced-to-string key
$dict[[string] $arr] # ditto, with a [string] cast

# Property syntax can be used too.
$dict."$arr" = 3
$dict."$arr"
```
